Add enhanced typings

as-com opened the pull request:
**Objective**
This PR adds enhanced typings for JSX as well as a number of other tweaks to enable Inferno to work with TypeScript's strict mode enabled.

Typings were copied from `@types/react` and bashed on until it and an example project compiled. Types and interfaces were modified to fit Inferno's internal types, classes were removed, and unimplemented features were commented out. Please feel free to rearrange/rewrite any of the type definitions I wrote if you have more knowledge of Inferno and/or TypeScript.

**Closes Issue**
It closes issue #686

Havunen commented on packages/inferno/src/core/component.ts:
```diff
@@ -177,5 +179,37 @@ export class Component<P, S> {
+export interface ComponentClass<P = {}> {
+  new (props?: P, context?: any): Component<P, {}>;
+
+  propTypes?: ValidationMap<P>;
```
Inferno does not have propTypes or contextTypes

Havunen commented on packages/inferno/src/core/implementation.ts:
```diff
@@ -25,28 +25,28 @@ export interface VNode {
   key?: any;
   className?: string;
-  [k: string]: any;
+  // [k: string]: any;
```
remove commented out code?

Havunen commented on packages/inferno/src/global.d.ts:
```diff
@@ -0,0 +1,177 @@
+React projects that don't include the DOM library need these interfaces to compile.
+React Native applications use React, but there is no DOM available. The JavaScript runtime
+is ES6/ES2015 only. These definitions allow such projects to compile with only `--lib ES6`.
+*/
```
Why do we need this for inferno?

as-com replied:
If I remove the reference to global, there are errors with HTMLWebViewElement. Also this would be needed for projects that use Inferno on the server side.

Havunen commented on packages/inferno/src/index.ts:
```diff
@@ -33,10 +42,3852 @@ if (process.env.NODE_ENV !== 'production') {
   }
 }
 
+
```
these type declarations could be in its own file ?

as-com commented:
FYI, I refactored the event typings to fit more with Inferno and rebased my PR on the current dev branch. 

I'm pretty sure there will be more cleanup/simplification necessary before this can be merged. I'll keep working on this maybe tomorrow.

Havunen commented:
There is issue related to typings: https://github.com/infernojs/inferno/issues/1295 , Could that be taken care of same time?

as-com commented:
Welp, looks like "tomorrow" turned into "next week..." :P

I think this PR is ready for a final review and merge.

LukeSheard commented on packages/inferno-compat/src/index.ts:
```diff
@@ -275,7 +275,7 @@ class PureComponent<P, S> extends Component<P, S> {
 class WrapperComponent<P, S> extends Component<P, S> {
```
We can neaten this up by having 

```
interface ContextProps {
    context: any;
}

type WrappedComponentProps<OriginalProps> = OriginalProps | ContextProps

class WrappedComponent<P, S> extends Component<WrappedComponentProps<P>, S> {
```

Havunen commented:
@as-com What is The status of this PR. We will need to release V5 because of breaking change in Webpack compatibility. Including this there would be nice

as-com commented:
@Havunen This PR is ready for merging, unless you find anything else that needs to be changed. 😄

Havunen commented:
How does this `global.d.ts` file work with our `index.d.ts` file? When you build the solution using `npm run build` it will auto generate index.d.ts file from source code and put that into dist folder. I'm not that familiar with typescript system, sorry if its silly question.

Havunen commented:
Could this issue be solved too in this PR? https://github.com/infernojs/inferno/issues/1209

as-com commented:
@Havunen #1209 should be fixed by this PR. 

On a second thought, the reference to global.d.ts may not be needed. I'll look into it later.

deamme commented on packages/inferno-router/src/utils.ts:
```diff
@@ -42,7 +42,7 @@ export function invariant(condition, format, a?, b?, c?, d?, e?, f?) {
 
 export const Children = {
-  forEach(children: any[], fn: Function): void {
+  forEach(children: any, fn: Function): void {
```
Typing seems to be more correct with `any[]` as before?

deamme commented:
You may need to run `npm run prettier:src:ts`.

Havunen commented:
Ok. I think this is good to go then! Lets merge it and go for v5!

Havunen commented:
Thank you for your work @as-com
